### PR TITLE
FAI-628: Refactor use of TypedValue/TypedVariable

### PR DESCRIPTION
**Pull request by manstis**
See https://issues.redhat.com/browse/FAI-628

Depends on https://github.com/kiegroup/kogito-runtimes/pull/1662

This has been tested with our integration tests and running within a _real_ environment.

I can provide instructions, @jiripetrlik @sjamboro, if needed to test this in a _real_ environment.

Many thanks for submitting your Pull Request :heart:! 

Please make sure that your PR meets the following requirements:

- [x] You have read the [contributors guide](https://github.com/kiegroup/kogito-runtimes#contributing-to-kogito)
- [x] Pull Request title is properly formatted: `KOGITO-XYZ Subject`
- [x] Pull Request title contains the target branch if not targeting main: `[0.9.x] KOGITO-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* <b>Pull Request</b>  
  Please add comment: <b>Jenkins retest this</b>
 
* <b>Quarkus LTS checks</b>  
  Please add comment: <b>Jenkins run LTS</b>

* <b>Native checks</b>  
  Please add comment: <b>Jenkins run native</b>

</details>

### CODE REVIEW

**Comment by manstis** [comment]
@kelvah Hello... FYI this PR changes the structure used for `Decision`s' inputs and outcomes and `Counterfactuals`.

They used to use a class solely in `trusty-service` and `trusty-storage` and, unfortunately (for me), the UI. We used to convert to/from them at the boundaries of `TraceEvent`->`trusty-service` and `trusty-service` <-> `explainability-service`. Now the same structures are used from `TraceEvent` through to `trusty-service`, `trusty-storage` and `explainability-service` (and will be the UI... once I've added a commit for those too....)

**Comment by manstis** [comment]
@sjamboro @jiripetrlik Please review your respective bits (UI, Java)

@danielezonca @r00ta @kelvah Please review your respective bits (UI, Java) too!

This PR is a _refactoring_ of data-structures... no new features.

File: trusty/trusty-service/trusty-service-common/src/main/java/org/kie/kogito/trusty/service/common/handlers/CounterfactualExplainerServiceHandler.java [hunk 1]
```diff
@@ -77,8 +79,16 @@ public CounterfactualExplainabilityResult explainabilityResultFrom(Counterfactua
                 dto.getStatusDetails(),
                 dto.isValid(),
                 stageFrom(dto.getStage()),
-                MessagingUtils.tracingTypedValueToModel(dto.getInputs()),
-                MessagingUtils.tracingTypedValueToModel(dto.getOutputs()));
+                dto.getInputs()
```
Comment (manstis) [inline trusty/trusty-service/trusty-service-common/src/main/java/org/kie/kogito/trusty/service/common/handlers/CounterfactualExplainerServiceHandler.java:81]: This conversion will be removed when [FAI-630](https://issues.redhat.com/browse/FAI-630) is completed.

File: trusty/trusty-service/trusty-service-common/src/main/java/org/kie/kogito/trusty/service/common/messaging/MessagingUtils.java [hunk 2]
```diff
@@ -52,55 +51,29 @@ private MessagingUtils() {
...
-    public static TypedValue modelToTracingTypedValue(TypedVariableWithValue value) {
+    public static CounterfactualSearchDomainDto modelToCounterfactualSearchDomainDto(CounterfactualSearchDomainValue value) {
         if (Objects.isNull(value)) {
             return null;
         }
         switch (value.getKind()) {
             case UNIT:
-                return new UnitValue(value.getTypeRef(), null, value.getValue());
+                CounterfactualSearchDomainUnitValue unit = value.toUnit();
+                return new CounterfactualSearchDomainUnitDto(unit.getType(), unit.isFixed(), MessagingUtils.modelToCounterfactualDomain(unit.getDomain()));
```
Comment (manstis) [inline trusty/trusty-service/trusty-service-common/src/main/java/org/kie/kogito/trusty/service/common/messaging/MessagingUtils.java:61]: These conversions will be removed when [FAI-629](https://issues.redhat.com/browse/FAI-629) is completed.

File: ui-packages/packages/trusty/src/components/Templates/Counterfactual/counterfactualReducer.ts [hunk 3]
```diff
@@ -213,13 +263,17 @@ const areRequiredParametersSet = (state: CFState): boolean => {
...
   // checking if all inputs have a domain specified, with the exception of
   // booleans (do not require one)
   return (
-    selectedInputs.length > 0 &&
-    selectedInputs.every(
-      input => input.domain || typeof input.value === 'boolean'
+    selectedInputValues.length > 0 &&
+    selectedInputValues.every(
+      inputValue =>
+        inputValue.domain || typeof inputValue.originalValue === 'boolean'
```
Comment (kelvah) [inline ui-packages/packages/trusty/src/components/Templates/Counterfactual/counterfactualReducer.ts]: You missed this one :slightly_smiling_face:
I think it should be something like:
```suggestion
        inputValue.domain || typeof inputValue.originalValue.value === 'boolean'
```

File: ui-packages/packages/trusty/src/components/Molecules/CounterfactualInputDomain/CounterfactualInputDomain.tsx [hunk 4]
```diff
@@ -18,21 +18,16 @@ const CounterfactualInputDomain = ({
       );
       break;
     case 'CATEGORICAL':
-      domain =
-        // special treatment for boolean values that have a 'fake' categorical
-        // domain
-        typeof input.value === 'boolean' ? (
-          <></>
-        ) : (
```
Comment (kelvah) [inline ui-packages/packages/trusty/src/components/Molecules/CounterfactualInputDomain/CounterfactualInputDomain.tsx]: I guess we do not need anymore this extra treatment for boolean domains. Is that so?

In that case I suppose you should have a look [at here inside CounterfactualTable.tsx](https://github.com/kiegroup/kogito-apps/blob/4cb0c8c1d714a9b49c73c4408b528519ed91dd64/ui-packages/packages/trusty/src/components/Organisms/CounterfactualTable/CounterfactualTable.tsx#L153) where the _fake_ categorical domain is created. (I cannot comment there because the code is unchanged. Thank you, github).

Right now the `dispatch` never fires because the if statement is looking for the old value structure. Could it be entirely removed, right?

**Review by kelvah — COMMENTED** [review]
Hi @manstis, I am still reviewing but I have a couple of things about CF to ask you.

File: ui-packages/packages/trusty/api-mock/mocks/modelData.js [hunk 5]
```diff
@@ -4,17 +4,17 @@ const DMN_1 = `
     <dmn:extensionElements/>
     <dmn:inputData id="_70F1A759-16EA-44EE-98C1-E9B3726382FF" name="Name">
       <dmn:extensionElements/>
-      <dmn:variable id="_544B578A-CB1D-4DE2-AB3C-22622689011A" name="Name" typeRef="string"/>
+      <dmn:variable id="_544B578A-CB1D-4DE2-AB3C-22622689011A" name="Name" type="string"/>
     </dmn:inputData>
     <dmn:decision id="_22CF5537-8C35-4EAB-B770-97BB5E58EEA8" name="Fred?">
       <dmn:extensionElements/>
-      <dmn:variable id="_5B829B9A-CA62-4174-B3A3-968EDF4F5C6D" name="Fred?" typeRef="boolean"/>
+      <dmn:variable id="_5B829B9A-CA62-4174-B3A3-968EDF4F5C6D" name="Fred?" type="boolean"/>
```
Comment (danielezonca) [inline ui-packages/packages/trusty/api-mock/mocks/modelData.js]: Is this change expected? As far as I see this is a DMN sample file

File: trusty/trusty-storage/trusty-storage-infinispan/src/test/java/org/kie/kogito/trusty/storage/infinispan/DecisionMarshallerTest.java [hunk 6]
```diff
@@ -35,11 +36,15 @@
...
     @Test
     public void testWriteAndRead() throws IOException {
-        List<DecisionInput> inputs = Collections.singletonList(new DecisionInput("id", "in", TypedVariableWithValue.buildUnit("nameIn", "number", JsonNodeFactory.instance.numberNode(10))));
+        List<DecisionInput> inputs = Collections.singletonList(new DecisionInput("id", "in", new UnitValue("nameIn", "number", JsonNodeFactory.instance.numberNode(10))));
         List<DecisionOutcome> outcomes = Collections.singletonList(new DecisionOutcome("id", "out",
                 DMNDecisionResult.DecisionEvaluationStatus.SUCCEEDED.toString(),
-                TypedVariableWithValue.buildUnit("nameOut", "number", JsonNodeFactory.instance.numberNode(10)),
-                Collections.singletonList(TypedVariableWithValue.buildUnit("nameOut", "number", JsonNodeFactory.instance.numberNode(10))),
+                new UnitValue("nameOut", "number", JsonNodeFactory.instance.numberNode(10)),
+                new HashMap<>() {
```
Comment (danielezonca) [inline trusty/trusty-storage/trusty-storage-infinispan/src/test/java/org/kie/kogito/trusty/storage/infinispan/DecisionMarshallerTest.java]: What about `Map.of` or `Collections.singletonMap`?

File: trusty/trusty-storage/trusty-storage-api/src/main/java/org/kie/kogito/trusty/storage/api/model/CounterfactualDomainRange.java [hunk 7]
```diff
@@ -29,18 +27,15 @@
     public static final String UPPER_BOUND = "upperBound";
 
     @JsonProperty(LOWER_BOUND)
-    @NotNull(message = "lowerBound object must be provided.")
     private JsonNode lowerBound;
 
     @JsonProperty(UPPER_BOUND)
-    @NotNull(message = "upperBound object must be provided.")
```
Comment (danielezonca) [inline trusty/trusty-storage/trusty-storage-api/src/main/java/org/kie/kogito/trusty/storage/api/model/CounterfactualDomainRange.java]: Why this change?

File: trusty/trusty-storage/trusty-storage-api/src/main/java/org/kie/kogito/trusty/storage/api/model/CounterfactualDomainCategorical.java [hunk 8]
```diff
@@ -29,13 +27,12 @@
     public static final String CATEGORIES = "categories";
 
     @JsonProperty(CATEGORIES)
-    @NotNull(message = "categories object must be provided.")
```
Comment (danielezonca) [inline trusty/trusty-storage/trusty-storage-api/src/main/java/org/kie/kogito/trusty/storage/api/model/CounterfactualDomainCategorical.java]: Same here

File: trusty/trusty-service/trusty-service-common/src/main/java/org/kie/kogito/trusty/service/common/CounterfactualParameterValidation.java [hunk 9]
```diff
@@ -16,94 +16,171 @@
...
+import org.kie.kogito.trusty.storage.api.model.DecisionOutcome;
 
 public class CounterfactualParameterValidation {
 
     private CounterfactualParameterValidation() {
         //Prevent instantiation of this utility class
     }
 
-    private interface Check<D extends TypedVariable<D>, C extends TypedVariable<C>> {
+    private interface Check<C, CV, D, DV> {
```
Comment (danielezonca) [inline trusty/trusty-service/trusty-service-common/src/main/java/org/kie/kogito/trusty/service/common/CounterfactualParameterValidation.java]: `CV` and `DV` are not used

File: trusty/trusty-service/trusty-service-common/src/main/java/org/kie/kogito/trusty/service/common/messaging/incoming/TraceEventConverter.java [hunk 10]
```diff
@@ -61,65 +56,17 @@ public static Decision toDecision(TraceEvent event, String sourceUrl, String ser
...
-                toMessageExceptionField(eventMessage.getException()));
-    }
-
-    public static MessageExceptionField toMessageExceptionField(org.kie.kogito.tracing.decision.event.message.MessageExceptionField eventException) {
-        return eventException == null
-                ? null
-                : new MessageExceptionField(eventException.getClassName(), eventException.getMessage(), toMessageExceptionField(eventException.getCause()));
+                eventOutput.getValue(),
+                eventOutput.getInputs() == null ? null : eventOutput.getInputs(),
+                eventOutput.getMessages() == null ? null : eventOutput.getMessages());
```
Comment (danielezonca) [inline trusty/trusty-service/trusty-service-common/src/main/java/org/kie/kogito/trusty/service/common/messaging/incoming/TraceEventConverter.java]: Are these null checks necessary? In both case it seems not necessary

File: trusty/trusty-storage/trusty-storage-api/src/main/java/org/kie/kogito/trusty/storage/api/model/CounterfactualSearchDomainCollectionValue.java [hunk 11]
```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates.
```
Comment (danielezonca) [inline trusty/trusty-storage/trusty-storage-api/src/main/java/org/kie/kogito/trusty/storage/api/model/CounterfactualSearchDomainCollectionValue.java]: Some license headers of new files are reporting 2020
```suggestion
 * Copyright 2021 Red Hat, Inc. and/or its affiliates.
```

File: trusty/trusty-storage/trusty-storage-api/src/main/java/org/kie/kogito/trusty/storage/api/model/CounterfactualGoal.java [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+ */
+package org.kie.kogito.trusty.storage.api.model;
+
+import org.kie.kogito.tracing.typedvalue.TypedValue;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class CounterfactualGoal {
```
Comment (danielezonca) [inline trusty/trusty-storage/trusty-storage-api/src/main/java/org/kie/kogito/trusty/storage/api/model/CounterfactualGoal.java]: This class, `CounterfactualInput` and `CounterfactualOutcome` are now the same, do you expect them to become different in the future? If not, what about a single generic `CounterfactualData`?

File: ui-packages/packages/trusty/api-mock/mocks/modelData.js (hunk 5 above)
Comment (manstis) [reply ui-packages/packages/trusty/api-mock/mocks/modelData.js]: Crikey. No this is not expected (or desirable). Super spot!

Comment (manstis) [reply ui-packages/packages/trusty/api-mock/mocks/modelData.js]: Rolled back the change to this file (WebStorm's refactoring support is a bit _enthusiastic_!)

File: trusty/trusty-storage/trusty-storage-infinispan/src/test/java/org/kie/kogito/trusty/storage/infinispan/DecisionMarshallerTest.java (hunk 6 above)
Comment (manstis) [reply trusty/trusty-storage/trusty-storage-infinispan/src/test/java/org/kie/kogito/trusty/storage/infinispan/DecisionMarshallerTest.java]: Indeed, much nicer. Changed.

File: trusty/trusty-storage/trusty-storage-api/src/main/java/org/kie/kogito/trusty/storage/api/model/CounterfactualDomainRange.java (hunk 7 above)
Comment (manstis) [reply trusty/trusty-storage/trusty-storage-api/src/main/java/org/kie/kogito/trusty/storage/api/model/CounterfactualDomainRange.java]: Reverted.. No good reason.

File: trusty/trusty-storage/trusty-storage-api/src/main/java/org/kie/kogito/trusty/storage/api/model/CounterfactualDomainCategorical.java (hunk 8 above)
Comment (manstis) [reply trusty/trusty-storage/trusty-storage-api/src/main/java/org/kie/kogito/trusty/storage/api/model/CounterfactualDomainCategorical.java]: Reverted.. No good reason.

File: trusty/trusty-service/trusty-service-common/src/main/java/org/kie/kogito/trusty/service/common/CounterfactualParameterValidation.java (hunk 9 above)
Comment (manstis) [reply trusty/trusty-service/trusty-service-common/src/main/java/org/kie/kogito/trusty/service/common/CounterfactualParameterValidation.java]: Removed. I also added a JavaDoc explaining the generic parameters for future reference.

File: trusty/trusty-service/trusty-service-common/src/main/java/org/kie/kogito/trusty/service/common/messaging/incoming/TraceEventConverter.java (hunk 10 above)
Comment (manstis) [reply trusty/trusty-service/trusty-service-common/src/main/java/org/kie/kogito/trusty/service/common/messaging/incoming/TraceEventConverter.java]: Agreed. Removed. This exists in `main` but the collections, if non-null, are streamed and mapped.. obviously now the same structures are passed around more and there's no mapping the null check is redundant :+1:

File: trusty/trusty-storage/trusty-storage-api/src/main/java/org/kie/kogito/trusty/storage/api/model/CounterfactualSearchDomainCollectionValue.java (hunk 11 above)
Comment (manstis) [reply trusty/trusty-storage/trusty-storage-api/src/main/java/org/kie/kogito/trusty/storage/api/model/CounterfactualSearchDomainCollectionValue.java]: Updated.

File: trusty/trusty-storage/trusty-storage-api/src/main/java/org/kie/kogito/trusty/storage/api/model/CounterfactualGoal.java (hunk 12 above)
Comment (manstis) [reply trusty/trusty-storage/trusty-storage-api/src/main/java/org/kie/kogito/trusty/storage/api/model/CounterfactualGoal.java]: Fair point.. **IF** we were marshalling with XML it _might_ improve readability but since we use JSON exclusively the duplicated classes can easily be considered clutter. I've removed the duplicates and added a `NamedTypedValue` class (in keeping with the existing `TypedValue` class).

File: ui-packages/packages/trusty/src/components/Molecules/CounterfactualInputDomain/CounterfactualInputDomain.tsx (hunk 4 above)
Comment (manstis) [reply ui-packages/packages/trusty/src/components/Molecules/CounterfactualInputDomain/CounterfactualInputDomain.tsx]: I have reinstated the _special_ handling of `boolean` Search Domains :+1: Good spot. Thank-you.

I don't believe either `main` or my reinstatement are 100% correct.. If I am adding a `boolean` field as a Search Domain should the category be the logical opposite of the original value? i.e. if it was `true` then the Search Domain is `false` and if it was `false` then the Search Domain `true`?

WDYT?

File: ui-packages/packages/trusty/src/components/Templates/Counterfactual/counterfactualReducer.ts (hunk 3 above)
Comment (manstis) [reply ui-packages/packages/trusty/src/components/Templates/Counterfactual/counterfactualReducer.ts]: Changed.

File: ui-packages/packages/trusty/src/components/Molecules/CounterfactualInputDomain/CounterfactualInputDomain.tsx (hunk 4 above)
Comment (kelvah) [reply ui-packages/packages/trusty/src/components/Molecules/CounterfactualInputDomain/CounterfactualInputDomain.tsx]: And you're right. It should send the opposite value as domain. I think the analysis is still working because the explainability side is ignoring the search domain for boolean inputs. So it's more like a placeholder value to set the categorical domain.

We can address this in a separate ticket eventually.

**Review by kelvah — APPROVED** [review]
Thank you @manstis, the UI side looks good to me.

File: trusty/trusty-service/trusty-service-common/src/test/java/org/kie/kogito/trusty/service/common/api/ExplainabilityApiV1IT.java [hunk 13]
```diff
@@ -270,76 +272,72 @@ void testCounterfactualRequestWithStructuredModel() {
...
-        assertNotNull(domain1Child1.getDomain());
-        assertTrue(domain1Child1.getDomain() instanceof CounterfactualDomainCategorical);
-        CounterfactualDomainCategorical domain1Child1Def = (CounterfactualDomainCategorical) domain1Child1.getDomain();
+        assertEquals("tViolation", domain1.getValue().getType());
+        assertEquals(3, domain1.getValue().toStructure().getValue().size());
+
+        Iterator<Map.Entry<String, CounterfactualSearchDomainValue>> domain1ChildIterator = domain1.getValue().toStructure().getValue().entrySet().iterator();
+        Map.Entry<String, CounterfactualSearchDomainValue> domain1Child1 = domain1ChildIterator.next();
+        Map.Entry<String, CounterfactualSearchDomainValue> domain1Child2 = domain1ChildIterator.next();
+        Map.Entry<String, CounterfactualSearchDomainValue> domain1Child3 = domain1ChildIterator.next();
```
Comment (danielezonca) [inline trusty/trusty-service/trusty-service-common/src/test/java/org/kie/kogito/trusty/service/common/api/ExplainabilityApiV1IT.java:312]: It is not related to this PR because it was already there but there is no guarantee that the entrySet always returns the elements in the same order so it should be better to avoid to make assumptions on the order of the elements.
Can you please create a ticket to review this test and avoid a possible flaky test?

**Review by danielezonca — APPROVED** [review]
Great work, well done 👍

**Comment by danielezonca** [comment]
@cristianonicolai 
Test `ProcessDataIndexPostgreSqlIT` failed, it is probably a flaky test

https://github.com/kiegroup/kogito-apps/commit/1fd962a906a05f1189ec5234ab0f162b26a385fe/checks/4010894061/logs

File: trusty/trusty-service/trusty-service-common/src/test/java/org/kie/kogito/trusty/service/common/api/ExplainabilityApiV1IT.java (hunk 13 above)
Comment (manstis) [reply trusty/trusty-service/trusty-service-common/src/test/java/org/kie/kogito/trusty/service/common/api/ExplainabilityApiV1IT.java:312]: I've created tickets for a few items we've discovered during this refactoring: https://issues.redhat.com/browse/FAI-652

**Comment by manstis** [comment]
Jenkins run native

**Comment by manstis** [comment]
@danielezonca @r00ta FYI.. native tests are failing locally. I am investigating/fixing.

**Comment by manstis** [comment]
Jenkins run native

**Comment by manstis** [comment]
Jenkins run native

**Comment by manstis** [comment]
Jenkins retest this

**Comment by manstis** [comment]
Jenkins retest this

**Review by jiripetrlik — APPROVED** [review]

**Comment by manstis** [comment]
Jenkins run native

**Comment by manstis** [comment]
Jenkins retest this

**Comment by sjamboro** [comment]
@manstis I cannot build kogito-apps project.
```
Kogito Apps :: Trusty Storage API .................. FAILURE
[ERROR] /home/sjamboro/Documents/Repository/manstis/kogito-apps/trusty/trusty-storage/trusty-storage-api/src/main/java/org/kie/kogito/trusty/storage/api/model/CounterfactualSearchDomainValue.java:[18,41] cannot find symbol
[ERROR]   symbol:   class BaseTypedValue
[ERROR] /home/sjamboro/Documents/Repository/manstis/kogito-apps/trusty/trusty-storage/trusty-storage-api/src/main/java/org/kie/kogito/trusty/storage/api/model/CounterfactualSearchDomainCollectionValue.java:[37,15] cannot find symbol
[ERROR]   symbol:   variable Kind

```

**Comment by manstis** [comment]
> @manstis I cannot build kogito-apps project.
> 
> ```
> Kogito Apps :: Trusty Storage API .................. FAILURE
> [ERROR] /home/sjamboro/Documents/Repository/manstis/kogito-apps/trusty/trusty-storage/trusty-storage-api/src/main/java/org/kie/kogito/trusty/storage/api/model/CounterfactualSearchDomainValue.java:[18,41] cannot find symbol
> [ERROR]   symbol:   class BaseTypedValue
> [ERROR] /home/sjamboro/Documents/Repository/manstis/kogito-apps/trusty/trusty-storage/trusty-storage-api/src/main/java/org/kie/kogito/trusty/storage/api/model/CounterfactualSearchDomainCollectionValue.java:[37,15] cannot find symbol
> [ERROR]   symbol:   variable Kind
> ```

Hi @sjamboro.

You need to build kiegroup/kogito-runtimes#1662 first (I did say that in this PRs description).

**Comment by manstis** [comment]
Jenkins retest this

**Comment by nmirasch** [comment]
Jenkins retest this

**Comment by manstis** [comment]
Jenkins run native

**Comment by manstis** [comment]
Jenkins run native

File: ui-packages/packages/trusty/src/components/Molecules/CounterfactualInputDomain/CounterfactualInputDomain.tsx (hunk 4 above)
Comment (manstis) [reply ui-packages/packages/trusty/src/components/Molecules/CounterfactualInputDomain/CounterfactualInputDomain.tsx]: Ticket: https://issues.redhat.com/browse/FAI-667

**Comment by manstis** [comment]
Jenkins retest this

**Review by sjamboro — APPROVED** [review]
it looks good.
I run Audit UI with mocked data and I built kogito-runtime and kogito-app localy and I was able to perform [e2e tests](https://github.com/manstis/kogito-apps/tree/FAI-628/ui-packages/packages/trusty/it-tests) without any problem.

**Comment by manstis** [comment]
Native build failure tracked here: https://issues.redhat.com/browse/FAI-668
